aya-obj/src/obj.rs

```diff
@@ -712,10 +712,30 @@ impl Object {
         Ok(())
     }
 
-    fn parse_program(&self, section: &Section) -> Result<(Program, Function), ParseError> {
-        let prog_sec = ProgramSection::from_str(section.name)?;
-        let name = prog_sec.name().to_owned();
+    fn parse_programs(
+        &mut self,
+        section: &Section,
+        symbols: HashMap<String, Symbol>,
+    ) -> Result<(), ParseError> {
+        let progs = symbols
+            .iter()
+            .map(|(name, symbol)| self.parse_program(section, name, symbol))
+            .collect::<Result<Vec<(Program, Function)>, ParseError>>()?;
+        for (p, f) in progs {
+            let key = p.function_key();
+            self.programs.insert(f.name.clone(), p);
+            self.functions.insert(key, f);
+        }
+        Ok(())
+    }
 
+    fn parse_program(
+        &self,
+        section: &Section,
+        name: &str,
+        symbol: &Symbol,
+    ) -> Result<(Program, Function), ParseError> {
+        let prog_sec = ProgramSection::from_str(section.name)?;
```

Comment:
This is going to be the same for all programs in the section, so it could be
moved to the caller

Reply:
Done